File: Sources/WInstallerCore/DiskEnumerator.swift
import Foundation

/// Enumerates physical disks through `diskutil` structured (`-plist`) output and
/// maps them into the domain `USBDrive` model. Internal/system disks are marked
/// unsafe for erase and filtered out by default (`REQ-USB-002`, `REQ-USB-005`).
///
/// All parsing is exposed as pure static functions so it can be fixture-tested
/// without a real disk.
public struct DiskEnumerator: Sendable {
    public static let diskutil = "/usr/sbin/diskutil"

    private let runner: CommandRunning

    public init(runner: CommandRunning = ProcessCommandRunner()) {
        self.runner = runner
    }

    /// Removable, external drives that are safe to consider as erase targets.
    public func removableDrives() async throws -> [USBDrive] {
        let listResult = try await runner.run(
            PlannedCommand(executable: Self.diskutil, arguments: ["list", "-plist"], isDestructive: false),
            timeout: .metadata
        )
        let disks = try Self.parseDiskList(listResult.standardOutput)

        var drives: [USBDrive] = []
        for disk in disks {
            let infoResult = try? await runner.run(
                PlannedCommand(executable: Self.diskutil, arguments: ["info", "-plist", disk.identifier], isDestructive: false),
                timeout: .metadata
            )
            guard let infoResult,
                  let drive = try? Self.parseDriveInfo(infoResult.standardOutput, volumes: disk.volumes)
            else { continue }

            // Safety: never surface internal or fixed disks as erase targets.
            guard drive.isRemovable, !drive.isInternal else { continue }
            drives.append(drive)
        }
        return drives
    }

    /// Re-reads a single disk's identity immediately before a destructive step.
    /// Used by `OperationExecutor` to abort if the device changed.
    public func info(for identifier: String) async throws -> USBDrive {
        let result = try await runner.run(
            PlannedCommand(executable: Self.diskutil, arguments: ["info", "-plist", identifier], isDestructive: false),
            timeout: .metadata
        )
        return try Self.parseDriveInfo(result.standardOutput, volumes: [])
    }

    // MARK: - Pure parsing

    public struct ParsedDisk: Equatable, Sendable {
        public var identifier: String
        public var volumes: [String]
    }

    /// Extracts whole-disk identifiers and their volume names from
    /// `diskutil list -plist` output.
    public static func parseDiskList(_ data: Data) throws -> [ParsedDisk] {
        let decoded = try PropertyListDecoder().decode(DiskUtilList.self, from: data)
        return decoded.allDisksAndPartitions.map { entry in
            let partitionVolumes = (entry.partitions ?? []).compactMap(\.volumeName)
            let apfsVolumes = (entry.apfsVolumes ?? []).compactMap(\.volumeName)
            return ParsedDisk(
                identifier: entry.deviceIdentifier,
                volumes: partitionVolumes + apfsVolumes
            )
        }
    }

    /// Maps `diskutil info -plist <id>` output onto the domain `USBDrive`.
    public static func parseDriveInfo(_ data: Data, volumes: [String]) throws -> USBDrive {
        let info = try PropertyListDecoder().decode(DiskUtilInfo.self, from: data)

        let removable = (info.removable ?? false)
            || (info.removableMedia ?? false)
            || (info.removableMediaOrExternalDevice ?? false)
            || (info.ejectable ?? false)

        let size = info.size ?? info.totalSize ?? 0
        let displayName = firstNonEmpty(info.mediaName, info.ioRegistryEntryName, info.volumeName)
            ?? info.deviceIdentifier
        let mediaName = firstNonEmpty(info.mediaName, info.ioRegistryEntryName) ?? displayName

        let resolvedVolumes = volumes.isEmpty
            ? [info.volumeName].compactMap { $0 }.filter { !$0.isEmpty }
            : volumes

        return USBDrive(
            bsdIdentifier: info.deviceIdentifier,
            displayName: displayName,
            mediaName: mediaName,
            size: size,
            isRemovable: removable,
            isInternal: info.isInternal ?? false,
            connectionType: info.busProtocol ?? "Unknown",
            partitionScheme: partitionScheme(from: info.content),
            fileSystem: firstNonEmpty(info.filesystemName, info.filesystemType) ?? "Unformatted",
            volumes: resolvedVolumes
        )
    }

    static func partitionScheme(from content: String?) -> String {
        switch content {
        case "GUID_partition_scheme": "GPT"
        case "FDisk_partition_scheme": "MBR"
        case "Apple_partition_scheme": "APM"
        case .some(let value) where !value.isEmpty: value
        default: "Unknown"
        }
    }

    private static func firstNonEmpty(_ values: String?...) -> String? {
        for value in values {
            if let value, !value.trimmingCharacters(in: .whitespaces).isEmpty {
                return value
            }
        }
        return nil
    }
}

// MARK: - Decodable models for diskutil plist output

struct DiskUtilList: Decodable {
    var allDisksAndPartitions: [DiskListEntry]

    enum CodingKeys: String, CodingKey {
        case allDisksAndPartitions = "AllDisksAndPartitions"
    }
}

struct DiskListEntry: Decodable {
    var deviceIdentifier: String
    var size: Int64?
    var content: String?
    var partitions: [DiskListPartition]?
    var apfsVolumes: [DiskListPartition]?

    enum CodingKeys: String, CodingKey {
        case deviceIdentifier = "DeviceIdentifier"
        case size = "Size"
        case content = "Content"
        case partitions = "Partitions"
        case apfsVolumes = "APFSVolumes"
    }
}

struct DiskListPartition: Decodable {
    var deviceIdentifier: String?
    var volumeName: String?
    var mountPoint: String?

    enum CodingKeys: String, CodingKey {
        case deviceIdentifier = "DeviceIdentifier"
        case volumeName = "VolumeName"
        case mountPoint = "MountPoint"
    }
}

struct DiskUtilInfo: Decodable {
    var deviceIdentifier: String
    var mediaName: String?
    var ioRegistryEntryName: String?
    var size: Int64?
    var totalSize: Int64?
    var isInternal: Bool?
    var removable: Bool?
    var removableMedia: Bool?
    var removableMediaOrExternalDevice: Bool?
    var ejectable: Bool?
    var busProtocol: String?
    var content: String?
    var filesystemName: String?
    var filesystemType: String?
    var volumeName: String?

    enum CodingKeys: String, CodingKey {
        case deviceIdentifier = "DeviceIdentifier"
        case mediaName = "MediaName"
        case ioRegistryEntryName = "IORegistryEntryName"
        case size = "Size"
        case totalSize = "TotalSize"
        case isInternal = "Internal"
        case removable = "Removable"
        case removableMedia = "RemovableMedia"
        case removableMediaOrExternalDevice = "RemovableMediaOrExternalDevice"
        case ejectable = "Ejectable"
        case busProtocol = "BusProtocol"
        case content = "Content"
        case filesystemName = "FilesystemName"
        case filesystemType = "FilesystemType"
        case volumeName = "VolumeName"
    }
}

File: apps/windows/WInstaller.Core/DiskEnumerator.cs
using System.Text.Json;

namespace WInstaller.Core;

/// <summary>
/// Enumerates physical disks through the bundled <c>list-disks.ps1</c> /
/// <c>disk-info.ps1</c> helpers (Get-Disk + Get-Partition as JSON) and maps
/// them onto the domain <see cref="UsbDrive"/> model. System/boot disks are
/// marked unsafe for erase and filtered out by default (REQ-USB-002,
/// REQ-USB-005).
///
/// All parsing is exposed as pure static functions so it can be fixture-tested
/// without a real disk.
/// </summary>
public sealed class DiskEnumerator
{
    private readonly ICommandRunner _runner;

    public DiskEnumerator(ICommandRunner? runner = null)
    {
        _runner = runner ?? new ProcessCommandRunner();
    }

    /// <summary>Removable, non-system drives that are safe to consider as erase targets.</summary>
    public async Task<IReadOnlyList<UsbDrive>> RemovableDrivesAsync(CancellationToken cancellationToken = default)
    {
        var result = await _runner.RunAsync(
            WindowsCommands.Script("list-disks.ps1", [], isDestructive: false),
            CommandTimeout.Metadata,
            cancellationToken).ConfigureAwait(false);

        if (!result.Succeeded)
        {
            return [];
        }

        return ParseDiskList(result.StandardOutput)
            .Where(drive => drive.IsRemovable && !drive.IsSystemDisk)
            .ToList();
    }

    /// <summary>
    /// Re-reads a single disk's identity immediately before a destructive step.
    /// Used by <see cref="OperationExecutor"/> to abort if the device changed.
    /// </summary>
    public async Task<UsbDrive> InfoAsync(int diskNumber, CancellationToken cancellationToken = default)
    {
        var result = await _runner.RunAsync(
            WindowsCommands.Script(
                "disk-info.ps1",
                ["-DiskNumber", diskNumber.ToString(System.Globalization.CultureInfo.InvariantCulture)],
                isDestructive: false),
            CommandTimeout.Metadata,
            cancellationToken).ConfigureAwait(false);

        if (!result.Succeeded)
        {
            throw new CommandRunnerException(
                CommandRunnerErrorKind.LaunchFailed,
                $"Could not re-read disk {diskNumber}: {result.StandardError}");
        }

        return ParseDriveInfo(result.StandardOutput);
    }

    // MARK: Pure parsing

    /// <summary>Parses the JSON array produced by <c>list-disks.ps1</c>.</summary>
    public static IReadOnlyList<UsbDrive> ParseDiskList(string json)
    {
        using var document = JsonDocument.Parse(json);
        var root = document.RootElement;

        // ConvertTo-Json collapses single-element arrays to a bare object.
        if (root.ValueKind == JsonValueKind.Object)
        {
            return [ParseDisk(root)];
        }
        if (root.ValueKind != JsonValueKind.Array)
        {
            return [];
        }
        return root.EnumerateArray().Select(ParseDisk).ToList();
    }

    /// <summary>Parses the single JSON object produced by <c>disk-info.ps1</c>.</summary>
    public static UsbDrive ParseDriveInfo(string json)
    {
        using var document = JsonDocument.Parse(json);
        return ParseDisk(document.RootElement);
    }

    private static UsbDrive ParseDisk(JsonElement element)
    {
        var busType = GetString(element, "BusType") ?? "Unknown";
        var isRemovableMedia = GetBool(element, "IsRemovableMedia");
        var isBoot = GetBool(element, "IsBoot");
        var isSystem = GetBool(element, "IsSystem");

        var volumes = new List<string>();
        string? fileSystem = null;
        if (element.TryGetProperty("Volumes", out var volumesElement) && volumesElement.ValueKind == JsonValueKind.Array)
        {
            foreach (var volume in volumesElement.EnumerateArray())
            {
                var label = GetString(volume, "FileSystemLabel");
                var letter = GetString(volume, "DriveLetter");
                fileSystem ??= GetString(volume, "FileSystem");
                var name = !string.IsNullOrWhiteSpace(label) ? label!
                    : !string.IsNullOrWhiteSpace(letter) ? $"{letter}:" : null;
                if (name is not null)
                {
                    volumes.Add(name);
                }
            }
        }

        var friendlyName = FirstNonEmpty(
            GetString(element, "FriendlyName"),
            GetString(element, "Model"));
        var diskNumber = GetInt(element, "Number") ?? -1;

        return new UsbDrive(
            DiskNumber: diskNumber,
            DisplayName: friendlyName ?? $"Disk {diskNumber}",
            MediaName: friendlyName ?? $"Disk {diskNumber}",
            Size: GetLong(element, "Size") ?? 0,
            // The Windows storage stack reports USB thumb drives as
            // BusType=USB; card readers additionally set IsRemovableMedia.
            IsRemovable: string.Equals(busType, "USB", StringComparison.OrdinalIgnoreCase) || isRemovableMedia,
            IsSystemDisk: isBoot || isSystem,
            ConnectionType: busType,
            PartitionScheme: PartitionScheme(GetString(element, "PartitionStyle")),
            FileSystem: FirstNonEmpty(fileSystem) ?? "Unformatted",
            Volumes: volumes);
    }

    internal static string PartitionScheme(string? style) => style switch
    {
        "GPT" => "GPT",
        "MBR" => "MBR",
        "RAW" => "Unformatted",
        var value when !string.IsNullOrWhiteSpace(value) => value!,
        _ => "Unknown",
    };

    private static string? FirstNonEmpty(params string?[] values) =>
        values.FirstOrDefault(value => !string.IsNullOrWhiteSpace(value));

    private static string? GetString(JsonElement element, string name) =>
        element.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.String
            ? value.GetString()
            : null;

    private static bool GetBool(JsonElement element, string name) =>
        element.TryGetProperty(name, out var value) && value.ValueKind is JsonValueKind.True;

    private static int? GetInt(JsonElement element, string name) =>
        element.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.Number
            ? value.GetInt32()
            : null;

    private static long? GetLong(JsonElement element, string name) =>
        element.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.Number
            ? value.GetInt64()
            : null;
}
